Added more descriptive error to run(``) (Fixes #19094)

SaschaMann opened the pull request:
As discussed in #19094 This PR adds a more descriptive error message when `run()` is called with an empty command:

``` julia
julia> run(``)
ERROR: ArgumentError: Empty command cannot be run
 in run(::Cmd) at ./process.jl:651
```

SaschaMann committed:
Added more descriptive error to catch run() (Fixes #19094)

StefanKarpinski commented on base/process.jl:
```diff
@@ -647,6 +647,9 @@ Run a command object, constructed with backticks. Throws an error if anything go
 including the process exiting with a non-zero status.
 """
 function run(cmds::AbstractCmd, args...)
+    if isempty(cmds.exec) # Issue 19094
```
I would not put the issue number in a comment here – we do that in test code but not in base code.

yuyichao replied:
The check seems wrong given that the function accept not only `Cmd` objects.

SaschaMann replied:
Ah okay, I saw it in other files every now and then so I thought it'd be some undocumented convention. Removed it, should be explained by the message it throws

SaschaMann committed:
Changed the check to work for AbstractCmd

SaschaMann committed:
Added tests for run() exceptions

SaschaMann commented on base/process.jl:
```diff
@@ -647,11 +647,15 @@ Run a command object, constructed with backticks. Throws an error if anything go
+        ps = spawn(cmds, spawn_opts_inherit(args...)...)
+        success(ps) ? nothing : pipeline_error(ps)
+    catch exc
+        if isa(exc, BoundsError)
```
AFAIK This can only throw a BoundsError if cmds is empty, so this should be fine to check? Correct me if I'm wrong, Thanks!

tkelman replied:
It's possible that any of the other functions might happen to throw a BoundsError. Wouldn't it be better to check if `cmds` is empty, rather than using a try-catch?

StefanKarpinski replied:
I agree that this should not be done with try/catch. How about putting the check and error raising in the appropriate `spawn` method? That way you'll already be in the appropriate case based on the type of `cmds`.

SaschaMann replied:
Makes sense, I'll do that (tomorrow though, getting late here). Thanks for the help!

tkelman commented on test/process.jl:
```diff
@@ -0,0 +1,7 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+@testset "run(``) tests" begin
```
better to put these in an existing file (or, in a different PR, split an existing test file apart so it more closely matches the `base/` file names) - any new test file won't actually run unless it's added to `choosetests.jl`

SaschaMann replied:
Moved them to `spawn.jl` below the equality tests for `Cmd` and `AndCmds`.

ararslan commented on base/process.jl:
```diff
@@ -647,8 +647,15 @@ Run a command object, constructed with backticks. Throws an error if anything go
+        success(ps) ? nothing : pipeline_error(ps)
+    catch exc
+        if isa(exc, BoundsError)
+            throw(ArgumentError("Empty command cannot be run"))
```
I'd rephrase this slightly as "cannot run empty command" (with lowercase "c" in "cannot").

SaschaMann committed:
Moved check to spawn() and added more tests

SaschaMann commented:
The other spawn methods all call `spawn(::Cmd, ...)` so adding the check to only that one should be enough, I think.

Also added some more tests and changed the phrasing of the error message.

SaschaMann commented on base/process.jl:
```diff
@@ -647,15 +650,8 @@ Run a command object, constructed with backticks. Throws an error if anything go
-        rethrow()
-    end
+    ps = spawn(cmds, spawn_opts_inherit(args...)...)
+    success(ps) ? nothing : pipeline_error(ps)
```
This just restores it to what it was originally.

tkelman reviewed (approved):
lgtm (should be squashed on merge)

ararslan commented:
Looks good to me--great job here! @StefanKarpinski and @yuyichao, would you two mind taking another look, since your requests appear to have been addressed?
